include/gunrock/framework/operators/filter/predicated.hxx

```diff
@@ -82,46 +83,43 @@ void execute_mgpu(graph_t& G,
   input_ends[num_gpus - 1] = input->end();
   
   // Map
-  size_t new_sizes[num_gpus];
+  int new_sizes[num_gpus];
   for(int i = 0; i < num_gpus; i++) {
     auto ctx = context.get_context(i);
     cudaSetDevice(ctx->ordinal());
     
-    auto new_length = thrust::copy_if(
+    auto new_end = thrust::copy_if(
       thrust::cuda::par.on(ctx->stream()),
       input_begins[i],
       input_ends[i],
       output_begins[i],
       predicate
     );
     
-    new_sizes[i] = thrust::distance(output_begins[i], new_length);
+    new_sizes[i] = (int)thrust::distance(output_begins[i], new_end);
     cudaEventRecord(ctx->event(), ctx->stream());
   }
-  
+
   // Sync
   auto context0 = context.get_context(0);
   for(int i = 0; i < num_gpus; i++) {
     auto ctx = context.get_context(i);
     cudaStreamWaitEvent(context0->stream(), ctx->event(), 0);
   }
   
-  // Reduce
-  size_t offset = new_sizes[0];
-  for(int i = 1; i < num_gpus; i++) {
-    thrust::copy(
-      thrust::device,
-      output_begins[i],
-      output_begins[i] + new_sizes[i],
-      output_begins[0] + offset
-    );
-    
+  // Reduce (using `input` as temporary storage)
+  int offset = 0;
+  for(int i = 0; i < num_gpus; i++) {
+    thrust::copy_n(thrust::device, output_begins[i], new_sizes[i], input->begin() + offset);
     offset += new_sizes[i];
   }
-  output->resize(offset);
   
-  // Cleanup
   cudaSetDevice(context0->ordinal());
+  
+  // Copy back to `output` -- there must be a better way to do this w/ pointers -- @neoblizz?
+  thrust::copy_n(thrust::device, input->begin(), offset, output->begin());
```

Comment:
I am assuming `input` "holds" the correct data at this point? If so, we can just swap the buffers and it should have the correct result. I took this control (control of Enactor to swap the buffers) out of the internal operator interface, but you can use a layer above (so, filter.hxx) to do so.

See;   `E->swap_frontier_buffers();` at https://github.com/bkj/essentials/blob/6bdcb4d50b0bd3f534d377d340dfadd10593a464/include/gunrock/framework/operators/filter/filter.hxx#L63

you can use an `if` to control that some algorithms need the "swap" and others don't; another example; https://github.com/bkj/essentials/blob/6bdcb4d50b0bd3f534d377d340dfadd10593a464/include/gunrock/framework/operators/advance/advance.hxx#L106-L110